### PR TITLE
sof_cs42l42: support 98396 smart speaker

### PR DESCRIPTION
**Pull request by macchian**
ASoC: Intel: sof_maxim_common: move speakers dapm relevant settings

ASoC: Intel: boards: maxim common settings restructure
- new max_98373/max_98390_dai_link()/_codec_conf() functions for variants of sof machine drivers.
- new get_maxim_num_codec(): look up hid/uid and return numbers of first codec device present on board.

ASoC: Intel: sof_maxim_common: create present codec comp/conf in runtime
- unify 98373/98390 relevant dapm settings and extern.
- create maxim_spk_codec_init(): combine max_98373_spk_codec_init() and 98390_codec_spk_init().
- revised get_maxim_codec_num() to create present codec relevant information when device matched.

ASoC: Intel: sof_cs42l42: support max98396 smart amplifier
- new max_98396_dai_link and sof_max98396_codec_conf, max_98396_ops
- create and link sof-adl-max98396-cs42l42.tplg

### CODE REVIEW

File: sound/soc/intel/common/soc-acpi-intel-adl-match.c [hunk 1]
```diff
@@ -531,6 +531,7 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_adl_machines[] = {
 		.quirk_data = &adl_max98360a_amp,
 		.sof_tplg_filename = "sof-adl-max98360a-cs42l42.tplg",
 	},
+
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-adl-match.c]: don't add lines please

File: sound/soc/intel/boards/sof_cs42l42.c [hunk 2]
```diff
@@ -655,6 +655,7 @@ static int sof_audio_probe(struct platform_device *pdev)
 
 	if (sof_cs42l42_quirk & SOF_SPEAKER_AMP_PRESENT)
 		sof_audio_card_cs42l42.num_links++;
+
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: spurious line. This doesn't help IMHO.

Comment (plbossart) [inline sound/soc/intel/boards/sof_cs42l42.c]: commit message

s/restruct/restructure/

typo on functions

Missing Signed-off-by.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 3]
```diff
@@ -10,67 +10,80 @@
 #include <sound/soc-dapm.h>
 #include <uapi/sound/asound.h>
 #include "sof_maxim_common.h"
+#include <sound/pcm_params.h>
 
 #define MAX_98373_PIN_NAME 16
+#define MAXIM_HID_NUMS 3
+#define MAXIM_MAX_AMPS 4
+#define MAXIM_NUM_CODECS 2
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: what does this represent?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 4]
```diff
@@ -10,67 +10,80 @@
...
+
+static char maxim_codec_name[SND_ACPI_I2C_ID_LEN];
+static struct snd_soc_codec_conf maxim_codec_conf[MAXIM_MAX_AMPS];
+static struct snd_soc_dai_link_component maxim_components[MAXIM_MAX_AMPS];
+
+static const char * const maxim_name_prefixes[] = { "Right", "Left", "Tweeter Right", "Tweeter Left" };
+static unsigned int num_codecs = 2;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: not idea what this means.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 5]
```diff
@@ -10,67 +10,80 @@
...
+int maxim_trigger(struct snd_pcm_substream *substream, int cmd)
 {
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
 	struct snd_soc_dai *codec_dai;
 	struct snd_soc_dai *cpu_dai;
 	int j;
 	int ret = 0;
 
+	if (strcmp(maxim_components[0].dai_name, "max98373-aif1"))
+		return 0;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: this seems to NOT be generic code, so what was the point of merging the functions if now we have codec-specific stuff in the helpers?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 6]
```diff
@@ -105,213 +118,163 @@ int max_98373_trigger(struct snd_pcm_substream *substream, int cmd)
...
-{
-	card->codec_conf = max_98373_codec_conf;
-	card->num_configs = ARRAY_SIZE(max_98373_codec_conf);
-}
-EXPORT_SYMBOL_NS(max_98373_set_codec_conf, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+EXPORT_SYMBOL_NS(maxim_trigger, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
 /*
- * Maxim MAX98390
+ * 2/4 Maxim codecs
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: what does 2/4 mean?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 7]
```diff
@@ -105,213 +118,163 @@ int max_98373_trigger(struct snd_pcm_substream *substream, int cmd)
...
+	link->init = maxim_spk_codec_init;
+	link->ops = &maxim_ops;
+};
+EXPORT_SYMBOL_NS(maxim_dai_link, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+
+const struct snd_soc_ops maxim_ops = {
+	.hw_params = maxim_hw_params,
+	.trigger = maxim_trigger,
+};
+EXPORT_SYMBOL_NS(maxim_ops, SND_SOC_INTEL_SOF_MAXIM_COMMON);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: i give up, I don't know how you organized the changes but it's just not easy to review. Please try to organize things in smaller patches where we can see what changes you are making. This sort of big code reshuffling is error-prone and impossible to review.

**Review by plbossart — CHANGES_REQUESTED** [review]
this is too invasive, please rework the changes to make them smaller and reviewable.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 5 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: @plbossart , my beginning thoughts was reducing the .ops codes. For better maintenance, I should add codec_ops separately, any advices?

const struct snd_soc_ops max_98373_ops = {
	.hw_params = max98373_hw_params,
	.trigger = 98373_trigger,
};
const struct snd_soc_ops max98390_ops = {
	.hw_params = max98390_hw_params,
};
const struct snd_soc_ops max_98396_ops = {
	.hw_params = max98373_hw_params,
};

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 7 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: @plbossart, before I submit desired smaller patches, share my idea first.
Maybe I should separate this function relevance out as one commit for easier review?

The maxim_compute_codec_conf handles min 2 codecs(L/R, default ) and max 4 codecs(R/L/TR/TL).

Iti is scanning each on board codec/uid if found in MAXIM_HID[]/uid_strings[] and assigning corresponding dai_name from CODEC_DAI[].

for example:
2 codecs available cases:
dev_name, prefix, dai_name
ADS8396-00 - > "Right" , (max98396-aif1)
ADS8396-01 -> "Left"    , (max98396-aif1)

4 codecs available cases:
max98390-00 - > "Right", (max98390-aif1)
max98390-01 -> "Left"    , (max98390-aif1)
max98390-02 -> "Tweeter Right"  ,  (max98390-aif1)
max98390-03 -> "Tweeter Left"    ,  (max98390-aif1)

**Comment by plbossart** [comment]
sorry @macchian, I can't follow what you are asking or trying to do. This is a very busy period of the year so you want your PRs to be crystal-clear, or they will mechanically fall in the TODO list when the rest is completed. Thanks for your understanding.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 8]
```diff
@@ -124,6 +124,51 @@ struct snd_soc_ops max_98373_ops = {
...
+ * hid: hardware id of device. Current Maxim amps model support.
+ * amp_nums: return the codec numbers of first acpi device.
+ */
+static int get_maxim_num_codec(void)
+{
+	const char * const UID_STRINGS[] = { "0", "1", "2", "3" };
+	const char * const MAXIM_HID[] = {"MX98373", "MX98390"};
+
+	unsigned int uid, amp_nums = 0;
+	unsigned int hid, cur = 0;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: don't mix on the same lines variables that are initialized and variables that are not.
Better yet, my preference is one variable per line.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 9]
```diff
@@ -124,6 +124,51 @@ struct snd_soc_ops max_98373_ops = {
...
+ * Look up present on board Maxim codecs hid and uid.
+ *
+ * uid: unique id of device. 0,1,2,3 total maximum 4 codecs.
+ * hid: hardware id of device. Current Maxim amps model support.
+ * amp_nums: return the codec numbers of first acpi device.
+ */
+static int get_maxim_num_codec(void)
+{
+	const char * const UID_STRINGS[] = { "0", "1", "2", "3" };
+	const char * const MAXIM_HID[] = {"MX98373", "MX98390"};
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: static?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 10]
```diff
@@ -124,6 +124,51 @@ struct snd_soc_ops max_98373_ops = {
...
+
+	unsigned int uid, amp_nums = 0;
+	unsigned int hid, cur = 0;
+	struct acpi_device *adev;
+
+	/* iterate each uid if presents. */
+	for (uid = 0; uid < ARRAY_SIZE(UID_STRINGS); uid++) {
+		/* finding the first match codec HID */
+		for (hid = cur; hid < ARRAY_SIZE(MAXIM_HID); hid++) {
+			adev = acpi_dev_get_first_match_dev(MAXIM_HID[hid], UID_STRINGS[uid], -1);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: where is the acpi_put_device() done?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 11]
```diff
@@ -124,6 +124,51 @@ struct snd_soc_ops max_98373_ops = {
...
+
+		if (!adev)
+			goto invalid_nums;
+
+		amp_nums++;
+	}
+
+invalid_nums:
+	if (amp_nums != 2 && amp_nums != 4) {
+		pr_warn("Invalid number of amps found: %d, expected 2 or 4\n", amp_nums);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: if you have 2 codecs, shouldn't the uids be 0 and 1? 

I am really lost on what this function tries to accomplish.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 12]
```diff
@@ -124,6 +124,51 @@ struct snd_soc_ops max_98373_ops = {
 };
 EXPORT_SYMBOL_NS(max_98373_ops, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
+/*
+ * Look up present on board Maxim codecs hid and uid.
+ *
+ * uid: unique id of device. 0,1,2,3 total maximum 4 codecs.
+ * hid: hardware id of device. Current Maxim amps model support.
+ * amp_nums: return the codec numbers of first acpi device.
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: what does this mean?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 13]
```diff
@@ -765,18 +760,7 @@ static struct snd_soc_dai_link *sof_card_dai_links_create(struct device *dev,
...
-				links[id].num_codecs = ARRAY_SIZE(max_98390_4spk_components);
-			} else {
-				links[id].codecs = max_98390_components;
-				links[id].num_codecs = ARRAY_SIZE(max_98390_components);
-			}
-			links[id].init = max_98390_spk_codec_init;
-			links[id].ops = &max_98390_ops;
-			links[id].dpcm_capture = 1;
-
+			max_98390_dai_link(&links[id]);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c:773]: not ok, you are not initializing dpcm_capture in that function.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 14]
```diff
@@ -158,6 +152,21 @@ static int get_maxim_num_codec(void)
 		if (!adev)
 			goto invalid_nums;
 
+		/* get device info */
+		physdev = get_device(acpi_get_first_physical_node(adev));
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_maxim_common.c]: where is the put_device() ?

**Review by plbossart — CHANGES_REQUESTED** [review]
I didn't finish the review @macchian, too many issues with the code IMHO.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 12 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: @plbossart ,thanks for your time for reviewing.
my thoughts is:
1. enumerating each acpi UID.
2. find first matched device by hid and uid. for example: MX98373:00, MX98373:01...
3. ”cur“ store the current matched codec index.
4. amp number increase 1
5. repeat 1-4 until current matched device UID all retrieved.
6. return amp numbers.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 9 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: @plbossart , my bad, no need to static in kernel permanently. Will remove it.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 8 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: ok.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 11 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: if you have 2 codecs, shouldn't the uids be 0 and 1?
yes.
When get first codec matched for example: "MX98373:00", amp number increase 1. The variable "cur" store the current matched codec index.
Iterating next uid and found "MX98373:01" by acpi_dev_get_first_match_dev() and so on until max uid 4 match or not.

@plbossart , hope my explanation doesn't confuse you more. Feel free to comment if any advices.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 10 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: @plbossart. the purpose is getting the number of codecs at all.
Because max_98373/98390_dai_link(), sof_max98373/98390_codec_conf() did. So I didn't put into device here.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 4 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: @plbossart I rewrote it simply I thought. Please help review.

File: sound/soc/intel/boards/sof_maxim_common.c (hunk 3 above)
Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: I remain MAXIM_MAX_CODECS 4 only.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 15]
```diff
@@ -13,6 +13,16 @@
 
 #define MAX_98373_PIN_NAME 16
 
+static const struct snd_kcontrol_new maxim_kcontrols[] = {
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: Please describe the purpose of this commit in the commit message. People need to know why we need this commit.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 16]
```diff
@@ -124,6 +124,52 @@ struct snd_soc_ops max_98373_ops = {
...
+		/* not found: next device id. */
+		if (!adev)
+			continue;
+
+		/* found: retrieve how many present codec of matched device. */
+		while (codec_nums < ARRAY_SIZE(CODEC_NUMS)) {
+			adev = acpi_dev_get_first_match_dev(MAXIM_DEVICES[i],
+							    CODEC_NUMS[codec_nums], -1);
+			if (!adev)
+				goto checked_num_support;
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: CODEC_NUMS is actually CODEC_UID, right? This doesn't work if you have discontinuous UIDs, like uid = 0, 2.
You can use 
```
for (j = 0; J < ARRAY_SIZE(CODEC_NUMS); j++) {
        ...
        if (!adev)
                continue;
        ... 
}
```

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 17]
```diff
@@ -124,6 +124,52 @@ struct snd_soc_ops max_98373_ops = {
 };
 EXPORT_SYMBOL_NS(max_98373_ops, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
+/*
+ * Find present Maxim codecs on board
+ *
+ * amp_nums: return the codec numbers, present 2 or 4 codecs support.
+ */
+
+int get_maxim_num_codec(void)
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: How about passing device name ("MX98373" or "MX98390") into get_maxim_num_codec()? So that you don't need to check both "MX98373" and "MX98390"

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 18]
```diff
@@ -326,17 +381,28 @@ const struct snd_soc_ops max_98390_ops = {
...
 EXPORT_SYMBOL_NS(max_98390_ops, SND_SOC_INTEL_SOF_MAXIM_COMMON);
 
-void max_98390_set_codec_conf(struct snd_soc_card *card, int ch)
+void max_98390_dai_link(struct snd_soc_dai_link *link)
+{
+	link->num_codecs = get_maxim_num_codec();
+	if (link->num_codecs == 4)
+		link->codecs = max_98390_4spk_components;
+
+	link->codecs = max_98390_components;
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: Won't this overwrite `link->codecs = max_98390_4spk_components;` if link->num_codecs == 4?

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 19]
```diff
@@ -326,17 +381,28 @@ const struct snd_soc_ops max_98390_ops = {
...
-	if (ch == ARRAY_SIZE(max_98390_4spk_codec_conf)) {
+	if (get_maxim_num_codec() == 4) {
 		card->codec_conf = max_98390_4spk_codec_conf;
 		card->num_configs = ARRAY_SIZE(max_98390_4spk_codec_conf);
-	} else {
-		card->codec_conf = max_98390_codec_conf;
-		card->num_configs = ARRAY_SIZE(max_98390_codec_conf);
 	}
+	card->codec_conf = max_98390_codec_conf;
+	card->num_configs = ARRAY_SIZE(max_98390_codec_conf);
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: card->codec_conf will be always max_98390_codec_conf and never be max_98390_4spk_codec_conf.

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 20]
```diff
@@ -152,12 +198,21 @@ int max_98373_spk_codec_init(struct snd_soc_pcm_runtime *rtd)
...
+	link->num_codecs = get_maxim_num_codec();
+	link->init = max_98373_spk_codec_init;
+	link->ops = &max_98373_ops;
+}
+EXPORT_SYMBOL_NS(max_98373_dai_link, SND_SOC_INTEL_SOF_MAXIM_COMMON);
+
+void sof_max98373_codec_conf(struct snd_soc_card *card)
 {
 	card->codec_conf = max_98373_codec_conf;
 	card->num_configs = ARRAY_SIZE(max_98373_codec_conf);
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: It seems max98373 always has two amps. If so, why do we need to use `get_maxim_num_codec()` to get codec numbers? Why don't we just set `link->num_codecs = 2` in max_98373_dai_link()?
If `get_maxim_num_codec()` could be 4, shouldn't we use different card->codec_conf and card->num_configs, just like `sof_max98390_codec_conf`?

File: sound/soc/intel/boards/sof_sdw_max98373.c [hunk 21]
```diff
@@ -34,21 +24,19 @@ static int spk_init(struct snd_soc_pcm_runtime *rtd)
 	if (!card->components)
 		return -ENOMEM;
 
-	ret = snd_soc_add_card_controls(card, mx8373_controls,
-					ARRAY_SIZE(mx8373_controls));
+	ret = snd_soc_add_card_controls(card, maxim_kcontrols, 2);
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_sdw_max98373.c]: Why don't we use `ARRAY_SIZE(maxim_kcontrols)`? Isn't `ARRAY_SIZE(maxim_kcontrols)` = 2?

**Comment by bardliao** [comment]
@macchian Please describe WHY in each commit message. Otherwise, people won't understand what are you trying to do.

**Comment by bardliao** [comment]
@macchian Please address the failures on
https://github.com/thesofproject/linux/actions/runs/4571562491/jobs/8069982390?pr=4052
https://github.com/thesofproject/linux/actions/runs/4571562491/jobs/8069982283?pr=4052

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 22]
```diff
@@ -319,24 +365,36 @@ int max_98390_spk_codec_init(struct snd_soc_pcm_runtime *rtd)
...
+void max_98390_dai_link(struct snd_soc_dai_link *link)
 {
-	if (ch == ARRAY_SIZE(max_98390_4spk_codec_conf)) {
+	link->codecs = max_98390_components;
+	link->num_codecs = ARRAY_SIZE(max_98390_components);
+
+	if (get_num_codecs("MX98390") == 4) {
+		link->codecs = max_98390_4spk_components;
+		link->num_codecs = ARRAY_SIZE(max_98390_4spk_components);
+	}
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_maxim_common.c]: Can we do
```
link->codecs = max_98390_4spk_components;
link->num_codecs = get_num_codecs("MX98390")
```
?
And same as max_98390_4spk_codec_conf

Comment (macchian) [reply sound/soc/intel/boards/sof_maxim_common.c]: @bardliao , how about these?

        link->num_codecs = get_num_codecs("MX98390");
        if (link->num_codecs == 4)
                link->codecs = max_98390_4spk_components;
        else
                link->codecs = max_98390_components;

Comment (bardliao) [reply sound/soc/intel/boards/sof_maxim_common.c]: > @bardliao , how about these?
> 
> ```
>     link->num_codecs = get_num_codecs("MX98390");
>     if (link->num_codecs == 4)
>             link->codecs = max_98390_4spk_components;
>     else
>             link->codecs = max_98390_components;
> ```

Not necessary, the first 2 elements in max_98390_4spk_components are identical to max_98390_components. And only the first 2 elements will be used when link->num_codecs = 2.

File: sound/soc/intel/boards/sof_nau8825.c [hunk 23]
```diff
@@ -212,8 +212,6 @@ static int sof_card_late_probe(struct snd_soc_card *card)
 static const struct snd_kcontrol_new sof_controls[] = {
 	SOC_DAPM_PIN_SWITCH("Headphone Jack"),
 	SOC_DAPM_PIN_SWITCH("Headset Mic"),
-	SOC_DAPM_PIN_SWITCH("Left Spk"),
-	SOC_DAPM_PIN_SWITCH("Right Spk"),
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_nau8825.c:216]: why is this removed without any replacement? Not following what you are trying to do anyways with this patch

File: sound/soc/intel/boards/sof_rt5682.c [hunk 24]
```diff
@@ -58,7 +58,6 @@
 #define SOF_SSP_BT_OFFLOAD_PRESENT		BIT(22)
 #define SOF_RT5682S_HEADPHONE_CODEC_PRESENT	BIT(23)
 #define SOF_MAX98390_SPEAKER_AMP_PRESENT	BIT(24)
-#define SOF_MAX98390_TWEETER_SPEAKER_PRESENT	BIT(25)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_rt5682.c:61]: why is this changed?

You are doing too many changes in this patch, it's impossible to review.

If you move stuff around, just move stuff and don't mix code moves with functionality changes.

**Review by plbossart — CHANGES_REQUESTED** [review]
This PR is way too invasive and poorly structured, impossible to review.

NAK. 

Please simplify this and make the code changes idiot-proof.

**Comment by macchian** [comment]
> This PR is way too invasive and poorly structured, impossible to review.
> 
> NAK.
> 
> Please simplify this and make the code changes idiot-proof.

@plbossart, @bardliao  I think let me just submit a single patch for 98396 enabling alone at this monument. Drop other series changes before I'm not coming up with clearer patches.

**Comment by plbossart** [comment]
I'll close this PR, @macchian if you want better support for 98396 please submit a new PR.
